geode-core/src/distributedTest/java/org/apache/geode/cache/client/ClientServerRegisterInterestsDUnitTest.java

```diff
@@ -224,27 +205,27 @@ public void testClientRegisterInterests() {
       Region<String, String> example = clientCache.getRegion(SEPARATOR + "Example");
 
       assertNotNull("'Example' Region in Client Cache was not found!", example);
-      assertEquals(1, example.size());
+      assertThat(example.size()).isEqualTo(1);
       assertTrue(example.containsKey("1"));
-      assertEquals("ONE", example.get("1"));
+      assertThat(example.get("1")).isEqualTo("ONE");
```

Comment:
This line and the 2 above it could be collapsed into:
```java
assertThat(example).containsOnly(entry("1", "ONE"));
```

Reply:
I think the existing line is clearer and the three lines make it easier to debug. I wonder if the error above would be clear enough.

Reply:
How so?
The prescribed method in a single like declares the map should only contain a single entry with the specified key and value. The error message here would be that either the map had other entries, and here are the entries, or it does not contain the 1 entry we asserted. 
As written, if the entry was missing at 210 the error would be "null" != "ONE". If there were more entries, 208 would error "8 != 1". Neither of these will mention anything about map size, entries, etc.

Reply:
introducing three checks into a single lines means each of those checks can assert.
 When only "ONE" is present  you get the following error.
 assertThat(example).containsOnly(entry("1", "TWO"));

java.lang.AssertionError: 
Expecting map:
  {"1"="ONE"}
to contain only:
  ["1"="TWO"]
map entries not found:
  ["1"="TWO"]
and map entries not expected:
  ["1"="ONE"]

Or you can get use 
assertThat(example.get("1")).isEqualTo("TWO");
and get the more intelligible 

org.opentest4j.AssertionFailedError: 
expected: "TWO"
 but was: "ONE"

Reply:
Perfectly good plan, thwarted by the reality of what is. This is the same reason that I think chaining is bad in tests (and generally code). It looks really good, but the error you get is not helpful, which is what you want especially in tests.

Reply:
Java should do the optimizations for you (which it does). It is no faster or slower to use chaining, vs not except.

Reply:
I much prefer the first form, describing the map, the expectation, the missing entries, and the unexpected entries. All of that information is very useful for diagnosing failures.

Adding `.as("region in client cache")` would make it even more helpful.

Reply:
The code below will produce one and only one failure per line and clicking the stack trace line will take you to the one failure
 

_assertThat(example).hasSize(1);_

entry count doesn't match ....

```
java.lang.AssertionError: 
Expected size: 1 but was: 2 in:
{"0"="ZERO", "1"="ONE"}

at org.apache.geode.cache.client.ClientServerRegisterInterestsDUnitTest.testClientRegisterInterests(ClientServerRegisterInterestsDUnitTest.java:211)
```


_assertThat(example).containsKey("1");_

key doesn't match

```
org.opentest4j.AssertionFailedError: 
expected: "ONE"
 but was: null
Expected :"ONE"
Actual   :null

at org.apache.geode.cache.client.ClientServerRegisterInterestsDUnitTest.testClientRegisterInterests(ClientServerRegisterInterestsDUnitTest.java:212)
```


_assertThat(example.get("1")).isEqualTo("ONE");_

value doesn't match 

```
org.opentest4j.AssertionFailedError: 
expected: "ONE"
 but was: "ZERO"
Expected :"ONE"
Actual   :"ZERO"
at org.apache.geode.cache.client.ClientServerRegisterInterestsDUnitTest.testClientRegisterInterests(ClientServerRegisterInterestsDUnitTest.java:213)
```

Any of the three failures will lead you to the one line below and you will have to break it apart  (if it was more complex) to test each condition.

_assertThat(example).containsOnly(entry("1", "ONE"));_


entry doesn't match ....

```
java.lang.AssertionError: 
Expecting map:
  {"2"="ONE"}
to contain only:
  ["1"="ONE"]
map entries not found:
  ["1"="ONE"]
and map entries not expected:
  ["2"="ONE"]

	at org.apache.geode.cache.client.ClientServerRegisterInterestsDUnitTest.testClientRegisterInterests(ClientServerRegisterInterestsDUnitTest.java:213)

```
count doesn't match ....

```
java.lang.AssertionError: 
Expecting map:
  {"0"="ZERO", "1"="ONE"}
to contain only:
  ["1"="ONE"]
but the following map entries were unexpected:
  ["0"="ZERO"]

	at org.apache.geode.cache.client.ClientServerRegisterInterestsDUnitTest.testClientRegisterInterests(ClientServerRegisterInterestsDUnitTest.java:213)
```

Key not present...
```
java.lang.AssertionError: 
Expecting map:
  {"0"="ZERO"}
to contain only:
  ["1"="ONE"]
map entries not found:
  ["1"="ONE"]
and map entries not expected:
  ["0"="ZERO"]

	at org.apache.geode.cache.client.ClientServerRegisterInterestsDUnitTest.testClientRegisterInterests(ClientServerRegisterInterestsDUnitTest.java:213)
```

Reply:
If this is preference, I prefer the original for my reasons above. If this is our coding standard, I will make the changes.

Reply:
I should add @demery-pivotal that I understand what you are saying, but I don't agree that you are losing all of the context. The context is there as soon as you click the stack trace. And better yet, there is better context because there is less out of context information on the line meaning the other assertions grouped in.

I think this discussion largely boils down to preference. I have accepted all of Jake's comments (all of which I agree with as they make the code better, with the exception of this one). The way I process information and debug, Jakes suggestion makes this code more complex than it needs to be.

Reply:
If I were diagnosing a failure of this test, I would want all of this important information:
- The entire content of the map.
- What the test expected about the content.
- Which expected entries were missing.
- Which existing entries were unexpected.

In this sequence of assertions:
```
assertThat(example.size()).isEqualTo(2);
assertThat(example.get("1")).isEqualTo("ONE");
assertTrue(example.containsKey("2"));
assertThat(example.get("2")).isEqualTo("TWO");
```
If the first assertion fails, I have NO way to know the content of the map.

If the second assertion fails, I have NO way to know either the key or the value of the other entry.

If the third assertion fails, the failure message tells me only that something wasn’t true. I have NO way to know either the key or the value of the other entry.

Some of the information I want is available not in the failure message, and not in the source code, but only in memory while the test is running.

If any of those first three assertions fails, the assertion exits the test, discarding useful diagnostic information without reporting it.

If I were diagnosing a failure of this test, I would quite annoyed to discover that not only did the author not give me the information I want, but actually insisted that I not have it.

The assertion that Jake and I are asking for gives me all of the information I want, and no unwanted information.

Reply:
This isn't just a preference but the prescribed way to use AssertJ.

Reply:
Ok, I will change it.